### PR TITLE
Localstore integration

### PR DESCRIPTION
**Pull request by janos**
This PR replaces LDBStore with new localstore DB in all parts of the code.

Work needed to be done is migration from legacy data store https://github.com/ethersphere/go-ethereum/pull/1297 and logging and tracing in localstore package. Beside this missing parts, this PR should provide a functional swarm node starting from a blank database.

### CODE REVIEW

**Comment by janos** [comment]
PR https://github.com/ethereum/go-ethereum/pull/19215 was opened by mistake on ethereum, which can be reused if this PR is accepted.

File: swarm/storage/localstore/mode_has.go
Comment (zelig) [inline swarm/storage/localstore/mode_has.go]: Get->Has

File: swarm/storage/localstore/localstore.go
Comment (zelig) [inline swarm/storage/localstore/localstore.go:32]: ?

**Review by zelig — COMMENTED** [review]
would it not make sense to split this PR and extract the Has/shed related changes?

**Comment by janos** [comment]
> would it not make sense to split this PR and extract the Has/shed related changes?

Yes, I will create a new PR, Has is pretty isolated change.

File: swarm/storage/localstore/localstore.go
Comment (janos) [reply swarm/storage/localstore/localstore.go:32]: A static check at compile time that DB is implementing all chunk store methods.

File: cmd/swarm/db.go
Comment (zelig) [inline cmd/swarm/db.go:170]: could this not be in the localstore pkg as one of the constructors?

File: swarm/network/stream/common_test.go
Comment (zelig) [inline swarm/network/stream/common_test.go]: should this not be in NetStore close?

File: swarm/network/stream/common_test.go
Comment (zelig) [inline swarm/network/stream/common_test.go:245]: why not return error here like on line 221

File: swarm/storage/localstore/subscription_pull.go
Comment (zelig) [inline swarm/storage/localstore/subscription_pull.go]: will be sent

File: swarm/network/stream/common_test.go
Comment (janos) [reply swarm/network/stream/common_test.go:245]: Yes, that would be better.

File: swarm/network/stream/common_test.go
Comment (janos) [reply swarm/network/stream/common_test.go]: No, localstoreCleanup() only deletes the temporary directory, nothing more.

File: cmd/swarm/db.go
Comment (janos) [reply cmd/swarm/db.go:170]: Actually, I am not sure that we need it. It only checks if goleveldb data is present in a directory. Is it really necessary to have that check for export and import?

File: swarm/network/stream/syncer.go
Comment (janos) [inline swarm/network/stream/syncer.go:121]: @zelig I added set sync call here opposed to change localstore.DB.Put to always add chunk to the garbage collection index. Adding that functionality beaks a lot of localstore tests that explicitly check that the chunk is not garbage collected until set sync is called in various scenarios.

In case that Put change is a preferred way, this is the required change:

```diff
diff --git a/swarm/storage/localstore/mode_put.go b/swarm/storage/localstore/mode_put.go
index ecad86dac..d6e982d62 100644
--- a/swarm/storage/localstore/mode_put.go
+++ b/swarm/storage/localstore/mode_put.go
@@ -147,6 +147,29 @@ func (db *DB) put(mode chunk.ModePut, item shed.Item) (err error) {
                return ErrInvalidMode
        }
 
+       // WARNING: This is a temporary approach to insert
+       // the chunk in GC index for all Put modes.
+       // The code under this warning, should be removed
+       // when SetModeSync is implemented in stream package.
+       if mode != chunk.ModePutRequest {
+               i, err := db.retrievalAccessIndex.Get(item)
+               switch err {
+               case nil:
+                       item.AccessTimestamp = i.AccessTimestamp
+                       db.gcIndex.DeleteInBatch(batch, item)
+                       gcSizeChange--
+               case leveldb.ErrNotFound:
+                       // the chunk is not accessed before
+               default:
+                       return err
+               }
+               item.AccessTimestamp = now()
+               db.retrievalAccessIndex.PutInBatch(batch, item)
+               db.gcIndex.PutInBatch(batch, item)
+               gcSizeChange++
+       }
+       // END WARNING
+
        err = db.incGCSizeInBatch(batch, gcSizeChange)
        if err != nil {
                return err
```

But localstore tests need to be changed and commented how to revert them later.

File: swarm/network/stream/common_test.go
Comment (zelig) [inline swarm/network/stream/common_test.go:183]: oops this was not closed?

File: swarm/network/stream/delivery.go
Comment (zelig) [inline swarm/network/stream/delivery.go:128]: this is supposed to be ModeSync no?
retrive -> retrieve

File: swarm/network/stream/delivery.go
Comment (zelig) [inline swarm/network/stream/delivery.go:270]: TODO: Enable this log line?

**Comment by nonsense** [comment]
Resolved conflicts triggered by my PR.

**Review by nonsense — APPROVED** [review]

**Review by acud — CHANGES_REQUESTED** [review]
Great work @janos. I pointed out a few issues on top of my traditional nitpicking <3

File: swarm/storage/localstore/localstore.go
Comment (acud) [inline swarm/storage/localstore/localstore.go]: @janos could you please review this delta? this is what I came up with for push tags. I actually was thinking it might be cleaner to just have the name of the index changed and leave all the rest to the PR that I'm working on since this is a bit of a wider change. We could, however, leave this as it is. I don't think it would break anything or cause any harm

Comment (janos) [reply swarm/storage/localstore/localstore.go]: @justelad I really like how you managed to make the most compact tags encoding. I had to validate it, and the result was TestDB_pushIndex_Tags that I pushed to this branch.

File: swarm/storage/localstore/export.go
Comment (janos) [reply swarm/storage/localstore/export.go]: I do not see how can we do reliable chunk assumptions. Maybe we can store the first file in tar that contains the exported `format` information, a some sort of version would be enough, to the "legacy" import is discovered by the import function, not provided by the user? For example `.swarm-export-format` file that contains export version (1 for the legacy, 2 for the new localstore) and if that (first) file is not present, assume legacy (1). Maybe we can be even more flexible for the future and have `.swarm-export.json` with a json encoded struct{Format:int}?

File: swarm/network/stream/common_test.go
Comment (janos) [reply swarm/network/stream/common_test.go:183]: Unfortunately, no.

File: swarm/network/stream/delivery.go
Comment (janos) [reply swarm/network/stream/delivery.go:128]: ModeGetSync is used in SwarmSyncerServer.GetData in handleWantedHashesMsg. Should SwarmChunkServer be used for requests only?

File: swarm/network/stream/delivery.go
Comment (janos) [reply swarm/network/stream/delivery.go:270]: This log line was removed on master. Do you think that it should be enabled again?

File: swarm/api/config_test.go
Comment (janos) [reply swarm/api/config_test.go:44]: I would not like to change this test in this PR. Maybe to create a new PR for this?

File: swarm/chunk/chunk.go
Comment (janos) [reply swarm/chunk/chunk.go:166]: Mode is not secondary, it is crucial for the operation and because other methods have mode, also, it is more similar to them, to have ctx and mode before parameters that differ in their signature. I think that this ordering makes the Store interface methods more similar.

File: swarm/network/stream/delivery.go
Comment (janos) [reply swarm/network/stream/delivery.go:239]: This is a delivery request. If we got the chunk from the node that is closer to the chunk, we do not need to put in the storage to sync it. I am not sure what is a `syncing message`? We are deciding here should we put this chunk in syncing index or not.

File: swarm/network/stream/stream.go
Comment (janos) [reply swarm/network/stream/stream.go:630]: We are differentiating in handleChunkDeliveryMsg function, just we do not need to assert type any more and one switch case is needed. Do you have any suggestion how to make this less crappy?

File: swarm/storage/localstore/export.go
Comment (janos) [reply swarm/storage/localstore/export.go]: I have implemented a backward compatible Import function with support for versioning https://github.com/ethersphere/go-ethereum/pull/1282/commits/2fe747e2191c4dec690d87c00d839670a1bbbdab. @justelad this can eliminate legacy flag in your PR.

**Comment by janos** [comment]
I am blocking this PR as it presents much more problems then improvements.

From gitter https://gitter.im/ethersphere/hq?at=5c9a376081b15c5e4b8ca43a:

> Elad @justelad 15:29
> right now i have made so little progress since localstore is, at the moment, some freak mutant hybrid of the old localstore and the new localstore

Please, do not review this PR until an alternative solution is presented.

File: swarm/network/stream/delivery.go
Comment (zelig) [reply swarm/network/stream/delivery.go:239]: sry i know i said this but it should only be the case for chunks outside our depth.
chunks within the area of responsibility should always sync

Comment (janos) [reply swarm/network/stream/delivery.go:239]: @zelig thanks for the explanation, I may have understood wrongly. Could you check if current code handles various cases correctly?

Comment (acud) [reply swarm/network/stream/delivery.go:239]: @zelig I was discussing this with @janos today and I have a few questions in this matter:
1. the chunk comes from somewhere, that somehwere should already be the so called "area of responsibility" (let's assume that we are talking about push syncing right now) - aren't we intervening in the process of syncing that a chunk should arrive correctly to the correct NN? also aren't we generating redundancy which is bigger than the premise of `redundance of degree 𝑛 in the network`? this is to say even more risky since area of responsibility is not symmetric between peers

2. I find it very hard to define a test case for this and so I would assume this is a premature optimisation. I think that right now with all that has to do with routing and syncing we should stick to correctness over performant, moreover, our code should be loyal to the spec we are writing. we shouldn't optimise anything at this point (IMO)

Comment (zelig) [reply swarm/network/stream/delivery.go:239]: @justelad 
1. no its not generating more redundancy the rest i dont understand
2. it is easy to construct test cases. this is about correctness

File: swarm/network/stream/delivery.go
Comment (zelig) [inline swarm/network/stream/delivery.go]:
````golang
if po >= depth || peerPO < po {
         mode = chunk.ModePutSync
} else {
         mode = chunk.ModePutRequest
}
```

**Review by zelig — CHANGES_REQUESTED** [review]

File: swarm/network/stream/delivery.go
Comment (janos) [reply swarm/network/stream/delivery.go]: Yes, this is more compact representation. I thought that the current one was easier to reason about, but this is good with a comment, as well.

Comment (zelig) [reply swarm/network/stream/delivery.go]: that is good but you got the comparison wrong, note
`po >= depth` -> sync

**Review by zelig — APPROVED** [review]

File: swarm/network/stream/delivery.go
Comment (janos) [reply swarm/network/stream/delivery.go]: Thanks Viktor.

**Review by acud — APPROVED** [review]

**Comment by janos** [comment]
Thanks for reviews. @holisticode I have set this PR ready for review, again. Before this gets merged, maybe we should decide about the migration PR that @justelad is working on. If this PR is merged, and released without the migration part, users will not have an option to keep their existing data.

File: swarm/storage/localstore/localstore.go
Comment (janos) [inline swarm/storage/localstore/localstore.go:289]: @zelig @justelad do we need to revert this index not to have Tags and remove tags from shed Item?

**Comment by janos** [comment]
@zelig @justelad I have added a check if chunk is already stored https://github.com/ethersphere/go-ethereum/pull/1282/commits/2e24e996320c1b7dafdf1abdc4fb053818a53f77

File: swarm/network/stream/common_test.go
Comment (holisticode) [inline swarm/network/stream/common_test.go:144]: why not `localStore.Close()` too?

File: swarm/network/stream/common_test.go
Comment (holisticode) [inline swarm/network/stream/common_test.go:185]: `localStore.Close()`?

Comment (janos) [reply swarm/network/stream/common_test.go:185]: netStore.Close() is calling localstore.Close()

File: swarm/network/stream/common_test.go
Comment (janos) [reply swarm/network/stream/common_test.go:144]: netStore.Close() is calling localstore.Close()

File: swarm/storage/localstore/localstore.go
Comment (justelad) [reply swarm/storage/localstore/localstore.go]: let's wait for the decision about the migration

Comment (acud) [reply swarm/storage/localstore/localstore.go:289]: @janos we definitely need a way to persist tags and I think that this is the correct place to do so. If anything, tags should be removed from `chunk.Chunk`

Comment (janos) [reply swarm/storage/localstore/localstore.go:289]: `chunk.Chunk` does not have reference to Tags, currently. Are you referring to `shed.Item` or some other code?

**Comment by nonsense** [comment]
@janos I think we can close this. I think we should consider this merged, since it is in the `swarm-rather-stable` branch, and we just need to decide on specific timing regarding releasing it.

**Comment by janos** [comment]
Yes, thanks, Anton. Closing this PR since it is merged into https://github.com/ethersphere/go-ethereum/tree/swarm-rather-stable branch.

**Pull request closed without merging**
